### ISSUE TITLE
Investigate use of docker

### ISSUE THREAD

**Issue by justinethier** [justinethier/cyclone#294]

**Comment by lassik** [comment]
We have made Docker containers for lots of Scheme implementations at [weinholt/scheme-docker](https://github.com/weinholt/scheme-docker). I tried making one for Cyclone but ran into some problems. Here's the `Dockerfile` I have so far:

```
# Need to use Debian sid because libck is not available in buster.
FROM debian:sid AS build
RUN apt-get update && apt-get -y --no-install-recommends install \
        build-essential \
        libck-dev \
        libtommath-dev \
        && rm -rf /var/lib/apt/lists/*

ADD https://github.com/justinethier/cyclone/archive/v0.11.2.tar.gz cyclone-0.11.2.tar.gz
RUN tar -xvf cyclone-0.11.2.tar.gz
WORKDIR cyclone-0.11.2
RUN make
RUN make install
RUN make test

FROM debian:sid
COPY --from=build /usr/local/ /usr/local/
```

**Comment by justinethier** [comment]
Unfortunately you cannot build cyclone directly using this repo, because it uses itself as part of the compilation process. You need to use the bootstrap repo. 

Here is a Dockerfile that should work:

```
# Need to use Debian sid because libck is not available in buster.
FROM debian:sid AS build
RUN apt-get update && apt-get -y --no-install-recommends install \
        build-essential \
        libck-dev \
        libtommath-dev \
        && rm -rf /var/lib/apt/lists/*

ADD https://github.com/justinethier/cyclone-bootstrap/archive/v0.11.3.tar.gz cyclone-bootstrap-0.11.3.tar.gz
RUN tar -xvf cyclone-bootstrap-0.11.3.tar.gz
WORKDIR cyclone-bootstrap-0.11.3
RUN make
RUN make install
RUN make test

FROM debian:sid
COPY --from=build /usr/local/ /usr/local/
```

**Comment by justinethier** [comment]
@lassik - Also, thanks for putting the Dockerfile together! See the above for a new one that should work...

**Comment by lassik** [comment]
No problem, it's fun :) With your fix to use `cyclone-bootstrap` instead of ordinary `cyclone` it worked easily. I thought I had done that but apparently made an oversight. I also had to install the runtime versions of libck and libtommath in the main container, naturally enough. Debian `sid-slim` images work just as well as `sid`.

**Comment by justinethier** [comment]
@lassik - This is great, thank you! 

Have you talked to @ecraven about any of this? He had mentioned using containers to standardize the installation of various schemes for the R7RS benchmarks, though all of that is currently run on Arch Linux. If nothing else, perhaps some of these docker files could be used as a starting point for creating Arch versions...

**Comment by justinethier** [comment]
@lassik - One more thing, does this Dockerfile persist the installations of libck and libtommath? They are required at runtime as well as to build. I ask because I fired this up on Debian 9.5 and got an error about libck not being found.

**Comment by lassik** [comment]
> Have you talked to @ecraven about any of this? He had mentioned using containers to standardize the installation of various schemes for the R7RS benchmarks, though all of that is currently run on Arch Linux. If nothing else, perhaps some of these docker files could be used as a starting point for creating Arch versions...

Neat! Nope, I had no idea he has done something like that. The present Dockerfiles started life as @weinholt's personal collection. He had done great groundwork so it was natural to contribute additions to it. If @ecraven has a collection ready, it would be great to merge ours with it. Perhaps we could start a `scheme-containers` GitHub organization. There could be one repo per implementation, and the implementation's author (or volunteers) could maintain the dockerfiles. I hear there are now other contenders for a container framework besides Docker so people who use those could contribute definitions for them as well.

**Comment by lassik** [comment]
> @lassik - One more thing, does this Dockerfile persist the installations of libck and libtommath? They are required at runtime as well as to build. I ask because I fired this up on Debian 9.5 and got an error about libck not being found.

Those should persist, and the container works on my computer. Do you have the latest Dockerfile from https://github.com/lassik/scheme-docker/blob/cyclone/cyclone/0.11.x/Dockerfile?

**Comment by justinethier** [comment]
@lassik - No, my understanding is that containers are a direction that he wanted to take, but there are no other containers available at this time. I was thinking that perhaps the Debian containers could be a starting point for Arch Linux containers - I assume a new set would need to be created for that platform?

Right, podman is the container technology that he mentioned. However, my understanding is that `podman` can be a drop-in replacement for docker (just alias `podman` instead of `docker`). So if we make a docker container then it will work with both technologies.

**Comment by justinethier** [comment]
@lassik - To answer your other comment, everything is probably fine, then. You don't happen to have libck/libtommath installed on the OS itself though, do you? 

Let me try again when I have a chance...

**Comment by lassik** [comment]
> I was thinking that perhaps the Debian containers could be a starting point for Arch Linux containers - I assume a new set would need to be created for that platform?

Yes, you have to create a new container for each Linux distro (and one for each different flavor of a particular Linux distro, if you want more than one flavor). But the Debian containers would indeed save him a ton of effort. Most of the commands are portable - things like running "make" or extracting a tar file.

I don't know whether the Dockerfile syntax allows easily re-using common fragments between container definitions. The syntax is very simple so if the definitions get unwieldy we can write them as S-expressions using macros trivially emit Dockerfiles in standard syntax.

If @weinholt is on board we can create the GitHub organization to collect all this stuff. @ecraven and others can join in and we can all work on the same repos.

> Right, podman is the container technology that he mentioned. However, my understanding is that `podman` can be a drop-in replacement for docker (just alias `podman` instead of `docker`). So if we make a docker container then it will work with both technologies.

Even better if the same dockerfiles can be re-used without changes. I've never heard of podman before. There's also [something called rkt](https://coreos.com/rkt/) as in "rocket" (not related to Racket). I'm not familiar with the details.

> To answer your other comment, everything is probably fine, then. You don't happen to have libck/libtommath installed on the OS itself though, do you?

It doesn't matter because the Linux distro in the container is completely isolated from the one on the host machine, so it can't access any libraries or other files from the host. Only the Linux kernel is shared. There are ways to mount host directories into a Docker container but it's usually not done, and none of our Dockerfiles do it. @weinholt can submit a Cyclone image to his Docker Hub account; then you should be able to start a Cyclone container by simply running `docker run -it weinholt/cyclone` without any preparations :)

**Comment by ecraven** [comment]
> If @weinholt is on board we can create the GitHub organization to collect all this stuff. @ecraven and others can join in and we can all work on the same repos.
> Even better if the same dockerfiles can be re-used without changes. I've never heard of podman before. There's also [something called rkt](https://coreos.com/rkt/) as in "rocket" (not related to Racket). I'm not familiar with the details.

> To answer your other comment, everything is probably fine, then. You don't happen to have libck/libtommath installed on the OS itself though, do you?

Just for info, I've idly started to create a container based on alpine
linux (very small size) for chibi-scheme (fully r7rs). My plan is to
create a container that I can pass an r7rs .scm file to, and that runs
that file. So all the mangling that needs to be done to get the
not-quite r7rs-Schemes to run r7rs code should be done *inside* the
container. I also want to support non-current versions of the Schemes,
so that older guile could be compared against newer guile, for example.

This might be subtly different from what you want to do, but if we
align, that'd be great!

Greetings, Peter

**Comment by lassik** [comment]
@ecraven thanks for stopping by! Your goals are almost the same as ours. Please take a look at our Chibi container. It almost does what you want, it's just based on Debian. We have one or two Alpine containers already.

I created an organization, https://github.com/scheme-containers and invited @ecraven and @weinholt. Feel free to create any repos you like. @ecraven, all our current work is the single repo https://github.com/weinholt/scheme-docker.
